core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -140,24 +150,54 @@ public Object apply(ServiceRequestContext ctx, HttpRequest req) {
 
     private CompletionStage<?> executeSyncOrAsync(ServiceRequestContext ctx, HttpRequest req) {
         final Object ret = apply(ctx, req);
-        return ret instanceof CompletionStage ?
-               (CompletionStage<?>) ret : CompletableFuture.completedFuture(ret);
+        if (ret instanceof CompletionStage) {
+            return ((CompletionStage<?>) ret).handle((obj, cause) -> {
+                if (cause == null) {
+                    return obj;
+                }
+                final HttpResponse response = getResponseForCause(ctx, req, cause);
+                if (response != null) {
+                    return response;
+                } else {
+                    return Exceptions.throwUnsafely(cause);
+                }
+            });
+        }
+        if (ret instanceof HttpResponse) {
+            return CompletableFuture.completedFuture(
+                    new ExceptionFilteredHttpResponse(ctx, req, (HttpResponse) ret));
+        }
+        return CompletableFuture.completedFuture(ret);
     }
 
     private Object invoke(ServiceRequestContext ctx, HttpRequest req, @Nullable AggregatedHttpMessage message) {
         try (SafeCloseable ignored = RequestContext.push(ctx, false)) {
             return method.invoke(object, parameterValues(ctx, req, parameters, message));
         } catch (Exception e) {
-            if (e instanceof InvocationTargetException) {
-                final Throwable cause = e.getCause();
-                if (cause != null) {
-                    return Exceptions.throwUnsafely(cause);
-                }
+            final Throwable cause = e instanceof InvocationTargetException ? firstNonNull(e.getCause(), e)
```

Comment:
This seems unnecessary because `getResponseForCause()` will call `Throwables.getRootCause()`.

Reply:
Ah, you're right. I added `Throwables.getRootCause()` at the last, so I missed removing this.